### PR TITLE
Modifying auto-bucket creation pattern.

### PR DESCRIPTION
**Pull request by andrew-glenn**
Part of the SigV2 -> SigV4 saga. This is necessary to facilitate SigV4 regional buckets. Initially, it'll create *more* buckets per test (1 per region), but those buckets will be retained (emptied if necessary). 

New buckets will be auto-created in this format:
`taskcat-<account_id_hash>-<region_name>`.

_Note: No modifications to `$[taskcat_autobucket]` are necessary due to the current design_

Special things to consider:
- This is opt-in via `s3_regional_buckets` from global down to test-level.
- Please pay special attention to these functions:

Dataclasses:
   - generate_bucket_name,
    - generate_regional_bucket_name,

Config:
  - _create_regional_bucket_obj
  - _create_legacy_bucket_obj

### CODE REVIEW

**Comment by andrew-glenn** [comment]
Closing as this should be in a major release, as it's a core change in behavior.

File: taskcat/_cli_modules/test.py [hunk 1]
```diff
@@ -118,7 +118,7 @@ def run(
             for test in buckets.values():
                 for bucket in test.values():
                     if bucket.name not in deleted:
-                        bucket.delete(delete_objects=True)
+                        bucket.empty()
```
Comment (andrew-glenn) [inline taskcat/_cli_modules/test.py]: Please pay special attention to this. Is this what we want to do?

Comment (avattathil) [reply taskcat/_cli_modules/test.py]: yeah this way we can keep using the same bucket sometimes like:

_account_id=hash(boto3.client('sts').get_caller_identity().get('Account'))
`taskcat-region-_account_id`

Comment (avattathil) [reply taskcat/_cli_modules/test.py]: looked close at the PR 
return f"{prefix}-{hashed_account_id}-{region_obj.name}"
^ works perfect

**Comment by andrew-glenn** [comment]
@avattathil 
@jaymccon 

This is ready for review. 

Note; I'm uncertain why code coverage isn't being picked up. :-/

**Comment by andrew-glenn** [comment]
Also: I launched the Linux Bastion across multiple accounts with this PR and achieved the desired behavior w/r/t buckets

File: taskcat/_dataclasses.py [hunk 2]
```diff
@@ -407,6 +421,9 @@ class ProjectConfig(JsonSchemaMixin, allow_additional_props=False):  # type: ign
     s3_bucket: Optional[S3BucketName] = field(
         default=None, metadata=METADATA["s3_bucket"]
     )
+    s3_regional_buckets: Optional[bool] = field(
```
[the remark this answers is not in the document]
Comment (andrew-glenn) [reply taskcat/_dataclasses.py]: My understanding.

Defaulting to false means that every value will be false. 

Global = False
Project = False
Test = False

So with that in mind, with Global or Project being true, but Test false, the net result is False due to inheritence. 

I additionally ran into issues getting actual/expected tests to pass without flipping this to an Optional value.

File: taskcat/_dataclasses.py [hunk 3]
```diff
@@ -429,7 +446,33 @@ class ProjectConfig(JsonSchemaMixin, allow_additional_props=False):  # type: ign
...
+
+def generate_bucket_name(project: str, prefix: str = "tcat"):
+    if len(prefix) > 8 or len(prefix) < 1:  # pylint: disable=len-as-condition
+        raise TaskCatException("prefix must be between 1 and 8 characters long")
+    alnum = string.ascii_lowercase + string.digits
+    suffix = "".join(random.choice(alnum) for i in range(8))  # nosec: B311
+    mid = f"-{project}-"
+    avail_len = 63 - len(mid)
+    mid = mid[:avail_len]
+    return f"{prefix}{mid}{suffix}"
```
[the remark this answers is not in the document]
Comment (andrew-glenn) [reply taskcat/_dataclasses.py]: Good callout. I'll move 'em .

File: taskcat/_dataclasses.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jaymccon) [reply taskcat/_dataclasses.py]: spot on, you know this code better than the author 😄

File: taskcat/_dataclasses.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (andrew-glenn) [reply taskcat/_dataclasses.py]: Actually, this generates a dependency loop

One of these depends on RegionObj, from dataclasses, so both end up require the other.

**Comment by andrew-glenn** [comment]
I've resolved the coverage issue locally; however I can't push to the `issue-455` branch as there's no delta between my local source and the remote branch. Closing this PR in favor of #504

**Pull request closed without merging**